### ISSUE TITLE
Explicit `const ref` intent on an owned class argument results in a segmentation fault

### ISSUE THREAD

**Issue by twesterhout** [chapel-lang/chapel#19128]
### Summary of Problem

The following piece of code segfaults (some previous iterations at obtaining the minimal example also produced deadlocks):

```chapel
use BlockDist;
use CPtr;

const OnePerLocale = LocaleSpace dmapped Block(LocaleSpace);

class BasisStates {
  var size : int;
  var representatives : [OnePerLocale] [0 ..# size] uint(64);
}

proc foo(const ref basisStates : BasisStates)
{
  coforall loc in Locales do on loc {
    writeln(c_ptrTo(basisStates.representatives[loc.id][0 ..# 1]));
  }
}

proc main()
{
  var representatives : [OnePerLocale] [0 ..# 13] uint(64) =
    [[31, 47, 55, 87, 91, 93, 103, 107, 155, 171, 173, 179, 341]:uint];
  var states = new BasisStates(13, representatives);
  foo(states);
}
```

* If `const ref basisStates` in the signature of `foo` is replaced with `basisStates`, things start to work.
* If `coforall` loop in `foo` is replaced by a `for` loop, things start to work.

There are two issues here. The important one is that the above code should not crash since passing owned classes by `const ref` is the default behavior, i.e. removing `const ref` should not change anything.

The less important issue (which I'm also less certain about) is that the above code should not compile at all... `c_ptrTo` expects a `ref`, but we pass it a `const ref`, don't we? If `const ref` is replaced by `ref` in the example, it still crashes, so the first issue still stands.

**Compile command:**

`chpl -o error error.chpl`

**Execution command:**

`./error -nl1`

### Configuration Information

- Output of `chpl --version`:
   ```
  chpl version 1.25.0
    built with LLVM version 11.1.0
   ```
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: unknown
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: smp *
  CHPL_GASNET_SEGMENT: fast
CHPL_TASKS: qthreads
CHPL_LAUNCHER: smp
CHPL_TIMERS: generic
CHPL_UNWIND: system *
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: system *
CHPL_AUX_FILESYS: none
```

Tagging @bradcray because of our discussion on gitter.

**Comment by bradcray** [comment]
@twesterhout : I haven't been able to reproduce this yet.  So far, I've tried:

* Chapel 1.25.1 and the current `main` branch on my Mac with LLVM enabled and CHPL_COMM=
  * none
  * gasnet/smp
  * gasnet/udp
* Chapel 1.25.0 on my Mac with the C back-end (I'm unable to build 1.25.0 on my Mac currently with LLVM enabled) w/ CHPL_COMM=
  * none
  * gasnet/smp
  * gasnet/udp

I'll try switching to linux64 next and see whether I get different results there.

**Comment by bradcray** [comment]
Aha, yes—linux64 made all the difference, without reverting to Chapel 1.25.0.  Looking more there now.

**Comment by bradcray** [comment]
Ah, and now that I've proven I can reproduce it, I'm seeing a use-after-free when running with address sanitizer on Mac as well, so that's reassuring (since it suggests that the behaviors aren't so different, just that I was getting "lucky" on Mac not to see a segfault).  I'm seeing the same use-after-free with valgrind.  So next step is to see what's going on there.

**Comment by bradcray** [comment]
(BTW, I marked this as a "gating issue" as I thought you said as much on gitter, but "things start to work." suggests that maybe you have a workaround?  Is that true, or does the workaround not apply in your original code?)

**Comment by bradcray** [comment]
Asan suggests to me that the `owned` class instance `basisStates` is being freed for the first time at the end of the `coforall` loop, and then a second time at the end of the program.  Adding a `with (const ref basisStates)` task intent to the `coforall` loop seems to be another way to work around the issue.

These observations, along with your comment that a normal `for` loop works, suggests to me that maybe something is going wrong with the shadow variables that the compiler inserts for `basisStates` by default.  That said, I'm not sure why using the explicit argument intent matters.  Here's an even simpler version of the program that exhibits the problem, at least under ASAN:

```chapel
class C {
  var x: int;
}

var myC = new C(42);

foo(myC);

proc foo(const ref myC: C) {
  coforall i in 1..10 do
    on Locales[i%numLocales] do
      writeln(i, " has ", myC);
}
```

I'm tagging @vasslitvinov on this as the expert on these task-private shadow variables and @mppf as the expert on owned classes and the interpretation of the blank argument intent.

Checking the generated code, in the breaking version, I'm seeing extra code like this before and after the loop:

```c
  _owned_C_chpl _cast_tmp__chpl;
  ...
  cast_tmp_chpl = borrow_chpl2(&_cast_tmp__chpl);
  // loop is here
  chpl__autoDestroy4(&_cast_tmp__chpl);
```

which I _think_ is being introduced in `adjustClassCastCall()` in functionResolution.cpp which Michael authored, so he may have better insight into what's going on.

**Comment by twesterhout** [comment]
> (BTW, I marked this as a "gating issue" as I thought you said as much on gitter, but "things start to work." suggests that maybe you have a workaround? Is that true, or does the workaround not apply in your original code?)

Thanks, this issue indeed has caused a bit of an interruption. When I say that things start to work I meant the example code which I provided, I have not properly tested the real code yet. I think there are also a few other workarounds I could try like replacing that owned class with a struct. However, since that project is a combination of multiple languages glued by FFI, it's quite tricky to judge whether it's my own bug or this issue...

I think what would greatly help if we could come up with a more robust workaround like forcing the compiler to not emit those free statements (I'd be happy to help out and play with it myself, but I'm afraid I don't know enough about the internals of Chapel). Robust would mean that valgrind & address sanitizer are happy rather than being simply lucky that the code does not segfault. Perhaps not specifying `const ref` argument intent or adding `with (const ref basisStates)` is already such a workaround, but that isn't quite clear to me from your comments. Do you still observe the faulty behavior with valgrind if you add any of those? (on my machine valgrind seems to be happy with your example if I remove `const ref` intent)

**Comment by mppf** [comment]
I've investigated a bit to try to answer "Why is the compiler adding this autoDestroy call"?

 * `_cast_tmp_` is an `owned C` (and not a reference to it) so addAutoDestroyCalls tries to deinit it
 * this variable has its DefExpr the preamble for the coforall loop
 * insertCasts adds the DefExpr to handle PRIM_MOVE myC_captemp, PRIM_DEREF arg myC, where myC_captemp has type C and myC has type _ref(_owned(C)). The PRIM_DEREF is what is causing a temporary value to appear here and I'm not sure it's necessary.
 * this PRIM_MOVE / PRIM_DEREF is added in https://github.com/chapel-lang/chapel/blob/main/compiler/resolution/functionResolution.cpp#L6152 . This code was added in PR #2062.
 
Simply removing the PRIM_DEREF gets the program to work. However I think it should probably be using PRIM_ASSIGN instead of PRIM_MOVE. I do not yet know that is sufficient all cases or if we need some more logic here to know when it should be added. 

<details>

```
diff --git a/compiler/resolution/functionResolution.cpp b/compiler/resolution/functionResolution.cpp
index bdff8bd109..98a6378f6f 100644
--- a/compiler/resolution/functionResolution.cpp
+++ b/compiler/resolution/functionResolution.cpp
@@ -6149,7 +6149,8 @@ static void captureTaskIntentValues(int        argNum,
 
       } else if (isReferenceType(varActual->type) ==  true &&
                  isReferenceType(capTemp->type)   == false) {
-        marker->insertBefore("'move'(%S,'deref'(%S))", capTemp, varActual);
+
+        marker->insertBefore(new CallExpr(PRIM_ASSIGN, capTemp, varActual));
 
       } else {
         marker->insertBefore("'move'(%S,%S)", capTemp, varActual);
```

</details>

**Comment by bradcray** [comment]
> it's quite tricky to judge whether it's my own bug or this issue...

Sorry if I hadn't made it clear.  Definitely a bug on our part (and I attached the 'bug' label to this issue once I'd determined that.

> Robust would mean that valgrind & address sanitizer are happy rather than being simply lucky that the code does not segfault. 

As some reassurance:  We test our suite of 13,000+ tests with valgrind and asan nightly, so you've presumably simply hit a case that our current test coverage doesn't hit.

> Perhaps not specifying const ref argument intent or adding with (const ref basisStates) is already such a workaround, but that isn't quite clear to me from your comments. Do you still observe the faulty behavior with valgrind if you add any of those? (on my machine valgrind seems to be happy with your example if I remove const ref intent)

Sorry to be unclear:  When I was calling those workarounds, I meant that they cause valgrind/asan to be happy.

But really, we need to fix this, so if you're willing to patch your version of 1.25, that'd be the lower overhead approach.  I'm running testing on @mppf's patch above to see what it does to the rest of the suite.

**Comment by bradcray** [comment]
@mppf / @twesterhout : I tested Michael's diff from the "details" section of his comment with CHPL_COMM=gasnet and valgrind testing without GASNet, and both passed, which I think is a good sign.  I'm going to try one more run with valgrind + --no-local for additional confidence (I wasn't seeing the segfault on --local compilations).

**Comment by mppf** [comment]
@bradcray - if something goes wrong with it, another option would be to use `PRIM_COERCE` instead of the `PRIM_CAST` in the event that the `PRIM_MOVE`'s LHS and RHS value types differ.

**Comment by bradcray** [comment]
@twesterhout : Just in case you didn't notice it (GitHub doesn't notify on these events, I think), @mppf filed his fix in PR #19139.  We typically don't merge PRs on Fridays unless we're planning on working over the weekend, so I'd expect this to be merged early next week.

While I'm very confident that this fixes the reduced test case you sent us, I'll be curious whether it also fixes your original program.  The patch is quite simple, so ought to be applicable to 1.25.x if you don't have a copy of the `main` development branch.

**Comment by bradcray** [comment]
(oh, and thanks for filing this!  that was a more subtle and worrisome bug than I've seen in awhile).

**Comment by twesterhout** [comment]
@bradcray and @mppf sorry for the delay (I got distracted by iQuHACK 2022 which took place this weekend). Thanks a lot for fixing this issue so quickly! I'll just switch to using the `main` branch. I'm pretty sure my original code should now also work, but if not, I'll open another issue :)

**Comment by bradcray** [comment]
@twesterhout : Hoping the lack of an update here means "everything worked great" or (more likely) "haven't had the chance to get back to it yet", but I was curious so thought I'd ping to see.

**Comment by twesterhout** [comment]
@bradcray my original piece of code does in fact work with Chapel built from the `main` branch. I've tried both with and without `const ref` annotations, both with and without concurrency (`forall`s), it all looks good. So I think this bug can be considered fixed!
